hyperstart/multicast.go

```diff
@@ -0,0 +1,158 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package hyperstart
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	hyper "github.com/hyperhq/runv/hyperstart/api/json"
+)
+
+type ctlDataType string
+
+const (
+	eventType ctlDataType = "ctlEvent"
+	replyType ctlDataType = "ctlReply"
+)
+
+type multicast struct {
+	reply []chan *hyper.DecodedMessage
+	event map[string]chan *hyper.DecodedMessage
+	ctl   net.Conn
+	sync.Mutex
+}
+
+func newMulticast(ctlConn net.Conn) *multicast {
+	return &multicast{
+		reply: []chan *hyper.DecodedMessage{},
+		event: make(map[string]chan *hyper.DecodedMessage),
+		ctl:   ctlConn,
+	}
+}
+
+func startCtlMonitor(ctlConn net.Conn) *multicast {
+	ctlMulticast := newMulticast(ctlConn)
+
+	go func() {
+		for {
+			msg, err := ReadCtlMessage(ctlMulticast.ctl)
+			if err != nil {
+				glog.Infof("Read on CTL channel ended: %s\n", err)
+				break
+			}
+
+			err = ctlMulticast.write(msg)
+			if err != nil {
+				glog.Errorf("Multicaster write error: %s\n", err)
+				break
+			}
+		}
+	}()
+
+	return ctlMulticast
+}
+
+func (m *multicast) buildEventID(containerID, processID string) string {
+	return fmt.Sprintf("%s-%s", containerID, processID)
+}
+
+func (m *multicast) sendEvent(msg *hyper.DecodedMessage) error {
+	var paeData hyper.ProcessAsyncEvent
+
+	err := json.Unmarshal(msg.Message, paeData)
+	if err != nil {
+		return err
+	}
+
+	uniqueID := m.buildEventID(paeData.Container, paeData.Process)
+	channel, exist := m.event[uniqueID]
+	if !exist {
+		return nil
+	}
+
+	channel <- msg
+
+	delete(m.event, uniqueID)
+
+	return nil
+}
+
+func (m *multicast) sendReply(msg *hyper.DecodedMessage) error {
+	var chanListLen int
+
+	m.Lock()
+	for {
+		chanListLen = len(m.reply)
+		if chanListLen > 0 {
+			break
+		}
+		m.Unlock()
+		time.Sleep(time.Millisecond)
+		m.Lock()
+	}
```

Comment:
Okay I think I see what you're explaining here. What do you think about sendReply() being called from write() and queueing the reply if needed (what you suggested). And another function called processQueueReply() called from listen().
So everything as you explained but split into 2 different functions for more clarity ?